sway-core/src/asm_generation/from_ir.rs

```diff
@@ -346,6 +367,77 @@ impl<'ir> AsmBuilder<'ir> {
         });
     }
 
+    // Read the argument(s) base from the script data.
+    fn read_args_value_from_script_data(&mut self, function: &Function, reg: &VirtualRegister) {
+        // Check if enough data was sent.
+        let data_len_reg = self.reg_seqr.next();
+        self.bytecode.push(Op {
+            opcode: either::Either::Left(VirtualOp::GTF(
+                data_len_reg.clone(),
+                VirtualRegister::Constant(ConstantRegister::Zero),
+                VirtualImmediate12 {
+                    value: GTFArgs::ScriptDataLength as u16,
+                },
+            )),
+            comment: "Get length of script data".into(),
+            owning_span: None,
+        });
+        let expected_data_len = function.args_iter(self.context).fold(0, |acc, arg| {
+            acc + ir_type_size_in_bytes(self.context, &arg.1.get_type(self.context).unwrap())
+        });
+        let expected_data_len_reg = self.reg_seqr.next();
+        self.bytecode.push(Op {
+            opcode: Either::Left(VirtualOp::MOVI(
+                expected_data_len_reg.clone(),
+                VirtualImmediate18 {
+                    value: expected_data_len as u32,
+                },
+            )),
+            comment: "Get expected length".into(),
+            owning_span: None,
+        });
+        let cmp_result_reg = self.reg_seqr.next();
+        self.bytecode.push(Op {
+            opcode: Either::Left(VirtualOp::LT(
+                cmp_result_reg.clone(),
+                data_len_reg,
+                expected_data_len_reg,
+            )),
+            comment: "Compare lengths".into(),
+            owning_span: None,
+        });
+        let label = self.reg_seqr.get_label();
+        self.bytecode.push(Op {
+            opcode: Either::Right(OrganizationalOp::JumpIfNotEq(
```

Comment:
It's unfortunate that we have to pay for this in the bytecode itself. What happens if we don't do these checks here?

Reply:
Nothing really. I was just following what's described in the issue. This whole length-checking feels like a crime against the gas golfing community :D

Reply:
This seems a little fragile anyways because script data could contain other things and this check may pass even in cases where the user has made a mistake. In other words, this check is necessary but not sufficient and paying gas for it, as you said, feels like a crime against the gas golfing community 😛 .

For users interacting with the script from the SDK, this shouldn't be a problem right?

Reply:
At this time, IDK, but we do generate ABIs and SDKs can check and warn if the user is not passing the right data. (So yes, we can easily handle this at the SDK level instead.)

ATM, we will need an escape hatch for variable length data though, which won't be a problem when we support `Vec` coding in SDKs.

Reply:
At this time, IDK, but we do generate ABIs and SDKs can check and warn if the user is not passing the right data. (So yes, we can easily handle this at the SDK level instead.)

ATM, we will need an escape hatch for variable length data though, which won't be a problem when we support `Vec` coding in SDKs.